Trucknavigation

OLFDB opened the pull request:
The navit routing engine is capable of truck navigation, taking into account sizes and weights (including axle weights) already.
This PR will enable maptool to provide the data needed by the routing engine in the map binfiles.

The truck profile in navit.xml was enhanced by supplying values for turn_around_penalty and turn_around_penalty2 to prevent uturns during routing.

OLFDB committed:
Add files via upload

Added Greek Keyboard Layout

OLFDB committed:
Add:gui/internal 

Added Greek keyboard

OLFDB committed:
Indent

OLFDB committed:
Indent

OLFDB committed:
Add files via upload

Indent

OLFDB committed:
Indent

OLFDB committed:
Indent

OLFDB committed:
Indent

OLFDB committed:
Indent

OLFDB committed:
FIX:DTD <navit> to not show errors in XML editors

OLFDB committed:
Merge pull request #1 from navit-gps/trunk

Update

OLFDB committed:
FIX: Removed duplicate comma

OLFDB committed:
FIX:WIN32 build with gtk_drawing_area

When building WIN32 with gtk_drawing_area the binfile map was not working (struct size of zipcd was 48)

OLFDB committed:
FIX:undefined symbol errors for cairo

OLFDB committed:
FIX:Undo changes

OLFDB committed:
FIX:Use cairo instead of pangocairo

OLFDB committed:
Feature: Support USERPROFILE folder for Windows

NAVIT_USER_DATADIR in Windows use USERPROFILE environment variable

OLFDB committed:
Merge pull request #2 from navit-gps/trunk

merge

OLFDB committed:
Indent

OLFDB committed:
indent

OLFDB committed:
indent

OLFDB committed:
indent

OLFDB committed:
indent

OLFDB committed:
indent

OLFDB committed:
Rework:USERPROFILE for Windows

Use extended environment_vars array for Windows. Use ~ to provide
homedir for Windows using USERPROFILE.

OLFDB committed:
FIX:build for macos

OLFDB committed:
Merge pull request #3 from navit-gps/trunk

Refactor:GUI/GTK:Adding pointer and label when displaying POI on map …

OLFDB committed:
Merge pull request #4 from navit-gps/trunk

Update from trunk

OLFDB committed:
FIX:Conditional apple build

OLFDB committed:
FIX:typo

lains committed:
Using HAVE_API_WIN32 instead of HAVE_API_WIN32_BASE (Both WinCE and Win32 have HAVE_API_WIN32_BASE defined)

OLFDB committed:
Update config.yml

Use latest stable git to support "update" command

OLFDB committed:
Revert "Update config.yml"

This reverts commit 9b403284f7da143838cc99e06748131e04616404.

OLFDB committed:
Revert "Update config.yml"

This reverts commit 9b403284f7da143838cc99e06748131e04616404.

wrong branch

OLFDB committed:
Merge branch 'trunk' of https://github.com/OLFDB/navit into trunk

OLFDB committed:
Revert "FIX:Use cairo instead of pangocairo"

This reverts commit 6b84e4796943cc960c739e5b066e667ed18b0b5d.

Wrong branch

OLFDB committed:
revert changes

Commit:
repair

Commit:
Revert "FIX: Removed duplicate comma"

This reverts commit 50eb35ea7d379d3dfaf539f927c20a1530ee828b.

Commit:
Revert "FIX:DTD <navit> to not show errors in XML editors"

This reverts commit e881f47c229a08f066f04d0b840251d2f114f55e.

Commit:
Revert "FIX:WIN32 build with gtk_drawing_area"

This reverts commit 43c649de9793331600ed4c1212c249f9da08622d.

OLFDB committed:
Disable sanity check for permanentrestrictions folder

pgrandin committed:
Merge branch 'trunk' into trunk

Commit:
Merge remote-tracking branch 'upstream/trunk' into trunk

Commit:
Minor revert on blank line

OLFDB committed:
Update update_download_center.sh

Changed NAVIT_DOWNLOAD_CENTER_REPO

OLFDB committed:
Update config.yml

OLFDB committed:
Update ci_sanity_checks.sh

OLFDB committed:
Merge pull request #10 from navit-gps/trunk

To trunk

OLFDB committed:
Update osm.c

Enable maptool to populate the data needed for trucknavigation.

OLFDB committed:
Update osm.c

Make maptool populate data needed for truck navigation.

OLFDB committed:
Disable debug output

OLFDB committed:
Update navit_shipped.xml

Added turn_around_penalty and turn_around_penalty2 to Truck profile.

OLFDB committed:
Merge branch 'trunk' into trucknavigation

jkoan commented on navit/maptool/osm.c:
```diff
@@ -597,6 +602,7 @@ static char *attrmap= {
     "w	boundary=civil		border_civil\n"
     "w	boundary=national_park	border_national_park\n"
     "w	boundary=political	border_political\n"
+    "?  boundary=low_emission_zone  low_emission_zone\n"
```
there should be a \t instead of two spaces

OLFDB committed:
Update osm.c

Format

metalstrolch commented on navit/maptool/osm.c:
```diff
@@ -597,6 +602,7 @@ static char *attrmap= {
     "w	boundary=civil		border_civil\n"
     "w	boundary=national_park	border_national_park\n"
     "w	boundary=political	border_political\n"
+    "?	boundary=low_emission_zone	low_emission_zone\n"
```
May be "w" in reality, as according to OSM wiki, boundary=low_emmision_zone is only allowed for areas and multi polygon relations which we map to polygons as well. This will result in quite big polygons on map probably. But it is very interesting information.

OLFDB replied:
I have not finished working on the low emission zones and this is another PR which should be removed from truck navigation. They are not showing up in the map. Maybe you have an idea why. I use

<itemgra item_types="low_emission_zone" order="0-">
    <polyline color="#00FF00" width="10" dash="5,5"/>
</itemgra>

in layout name="polygons". Changing to "w" type doesn't solve it.

metalstrolch replied:
I've committed the fix for low_emission_zone. Now you can have it in itemgra. However the combination of width != 1 and dash results in "funny" behaviour as graphics engine cannot dash wide lines due to the way they are drawn. I'd suggest transparent solid line.
To use it for routing however is a different story.

OLFDB replied:
Thank you, but it's still not showing up on the map.

metalstrolch replied:
OK, got it. low emission zones are mostly tagged on boundary relations, not on multipolygon ones. OSM wiki is not clear about this. Wil push fix. Caution: fix renames low_emission_zone to more correct poly_low_emission_zone. Adapt your layout accordingly

OLFDB replied:
Perfect. It works fine now. Thank you.

metalstrolch commented on navit/maptool/osm.c:
```diff
@@ -1103,6 +1119,41 @@ void osm_add_tag(char *k, char *v) {
             flags[0] |= AF_SPEED_LIMIT;
         level=5;
     }
+    if (!g_strcmp0(k, "maxheight")) {
```
Think we need to deal with imperial units as well. See https://wiki.openstreetmap.org/wiki/Key:maxheight
Same for the other dimensions.

metalstrolch commented on navit/maptool/osm.c:
```diff
@@ -53,6 +53,11 @@ static struct item_bin item;
 
 
 int maxspeed_attr_value;
+double maxheight_attr_value;
```
Why "double"? Since the units are converted to e.g. "cm" anyway, there's no need to have floats here. Later the attrs are added as int to the items anyway.

OLFDB replied:
My first guess was to use the osm values directly, but when I saw the unit used for the vehicle attributes are in cm I didn't change the type.

metalstrolch commented on navit/maptool/osm.c:
```diff
@@ -1103,6 +1119,41 @@ void osm_add_tag(char *k, char *v) {
+        level = 5;
+    }
+    if (!g_strcmp0(k, "maxlength")) {
+        maxlength_attr_value = (int)(atof(v)*100);
```
The "*100" is the reason there should be no float required.

metalstrolch reviewed (changes requested):
Good work. I'm looking forward for the low emmission zones, as those can be very interesting on map.

OLFDB committed:
Update osm.c

Using int variables and initialization of maxaxleload_attr_value

metalstrolch committed:
Add boundary=low_emission_zone to binfile map.

No it can be displayed on map if required. low emission zones are to be
mapped as areas. Add low_emission_zone to itemgras.

OLFDB committed:
Update osm.c

Imperial unit handling

OLFDB committed:
Merge branch 'trunk' into trucknavigation

metalstrolch committed:
Fix multi polygon detection

OLFDB committed:
Update navit_layout_car_shipped.xml

Added poly_low_emission_zone to polygons layer

OLFDB committed:
Update osm.c

Added weight conversion for values without unit identifier.
Disabled debug output.

OLFDB committed:
Update osm_xml.c

To be able to find items with wrong data for height, size, or weight.

OLFDB committed:
Update osm.c

To be able to find items with wrong height, size, weight.

OLFDB commented:
After converting some maps using maptool: how to handle e.g. below_default for the height? If we convert this height to 0 the way gets not flagged, but it will definitely lead to issues for trucks. We could use a default value of 250 so the way is excluded in truck routing.

OLFDB committed:
Update navit_layout_car_dark_shipped.xml

Added "poly_low_emission_zone"

OLFDB committed:
enhanced conversion functions

OLFDB committed:
Merge branch 'trucknavigation' of https://github.com/OLFDB/navit into trucknavigation

OLFDB committed:
handle onditional speed limits

OLFDB commented:
Added conditional speed limits. WIP.
![condspeedlimits](https://user-images.githubusercontent.com/5459286/114275067-6db35680-9a21-11eb-93f0-98520cea5222.jpg)

metalstrolch committed:
Fix:vehicle:geoclue Fix cmake usage (#1109)

OLFDB committed:
gui internal to change vehicle dimensions. Store dimensions in XML file. Improved layout conditional speed warner

OLFDB committed:
Merge branch 'trunk' into trucknavigation

OLFDB committed:
Merge branch 'trunk' into trucknavigation

OLFDB committed:
Merge branch 'trunk' into trucknavigation

OLFDB committed:
Update CMakeLists.txt

OLFDB committed:
Update

OLFDB committed:
Merge branch 'trucknavigation' of https://github.com/OLFDB/navit into trucknavigation

OLFDB committed:
Delete vehicleprofile.c

OLFDB committed:
added libxml2-dev

OLFDB committed:
update

OLFDB committed:
update

OLFDB committed:
update

OLFDB committed:
update

OLFDB committed:
update

OLFDB committed:
update

OLFDB committed:
added libxml2-dev

OLFDB committed:
added libxml2

OLFDB committed:
added libxml2

OLFDB committed:
added libxml2

OLFDB committed:
removed libxml2

OLFDB committed:
removed libxml2

OLFDB committed:
removed libxml2

OLFDB committed:
tmpfile not available in WINCE

OLFDB committed:
remove not available on WINCE

OLFDB committed:
Update

OLFDB committed:
Update

OLFDB committed:
Update

OLFDB committed:
Update

OLFDB committed:
Update

OLFDB committed:
Update for Android

OLFDB committed:
Update for Android

OLFDB committed:
Update for FDROID

OLFDB committed:
Update for FDROID

OLFDB committed:
Merge branch 'trunk' into trucknavigation

OLFDB committed:
improved speedwarners

OLFDB committed:
Merge branch 'trucknavigation' of https://github.com/OLFDB/navit into trucknavigation

OLFDB committed:
improved speedwarners

OLFDB committed:
rename

OLFDB committed:
minus is not allowed for android ressourec names

OLFDB committed:
create unscaled png images as well, use scaled image for conditional speed warner

OLFDB committed:
heightprofile shows heightmeters

OLFDB committed:
translation for vertical meter and distance

OLFDB commented:
Added vertical meters to the heightline widget in internal gui as that might be of interest planning a route with a truck, or bike, or pedestrian.
![heightlines](https://user-images.githubusercontent.com/5459286/117580607-06102a00-b0f9-11eb-8faf-0d7c0e04bdf5.jpg)

OLFDB committed:
emission class and hazmat value for vehicle

OLFDB commented:
Improvements for conditional speed limits. Backward limits are supported as well. Improved gui, now showing icons for danger, hazmat and wet condition. Hazmat conditions shown only if vehicle has hazmat set in it's profile. 

```
<osd enabled="yes" background_color="#FFFFFF00" color="#FFFFFFFF"
    type="cond_speed_warner"
   w="130" h="130" x="78%" y="-100%" font_size="250"
   speed_exceed_limit_offset="5" speed_exceed_limit_percent="10"              
   announce_on="1" label="images:maxspeed_48_48.png:maxspeed_48_48.png:empty.png:stvo_1040_30.png:stvo_1052_30.png:stvo_1053_35.png:stvo_1006_31:" name="cond_speed_warner" />
   <!--										 CONDITION        HAZARD	   WET		    DANGER         -->
```

![unfallgefahr](https://user-images.githubusercontent.com/5459286/118081716-39cea680-b3bc-11eb-9375-092bb30b299c.jpg)
![hazard](https://user-images.githubusercontent.com/5459286/118081718-3a673d00-b3bc-11eb-867a-a24d3fc5b778.jpg)
![wet](https://user-images.githubusercontent.com/5459286/118081719-3affd380-b3bc-11eb-9e36-fdd073cd9622.jpg)

Routing recognizing low emission zones

Thanks to mvglasow for his proposals for the algorithms to find intersections of a line with a polygon. I'm still trying to get routing to recognize the low emission zones. route value for segments in lez is INT_MAX which should result in blocked route if route ends in a lez. I have tested with a small map of Berlin only, so there is only one lez in the map.
I'm reading the lez polygon in route_graph_add_street and searching intersection with current segment in route_graph_add_segment. For detection I took the line_intersection function from gui_internal.c. It detects intersections, but a route is still calculated. 

Any hint would be appreciated.

OLFDB committed:
fix for loop initial declarations are only allowed in C99 mode

OLFDB committed:
rename of function line_intersection

OLFDB committed:
translation Höhenmeter

OLFDB committed:
Merge branch 'trucknavigation' of https://github.com/OLFDB/navit into trucknavigation

OLFDB committed:
icons for conditional speed warner

OLFDB committed:
Merge branch 'trucknavigation' of https://github.com/OLFDB/navit into trucknavigation

OLFDB committed:
store and restore active vehicleprofile in/from gui_internal.txt

OLFDB committed:
Merge branch 'trucknavigation' of https://github.com/OLFDB/navit into trucknavigation

OLFDB committed:
loglevel reduced

OLFDB committed:
disbale lez route value, hazmat and emission class selectable via gui

OLFDB committed:
routing takes low emission zones into account when configured in vehicle profile. Default is low emission zones are allowed. vehicle_gpsd shows reconnect errors only for the selected vehicle when more than one gpsd vehicle is configured

OLFDB committed:
use line_intersection from graphics

OLFDB commented:
Finally the routing taking into account the low emissions zones is working.
I think the PR is finished now.

https://user-images.githubusercontent.com/5459286/119258878-e46c7380-bbcb-11eb-9ffb-0385843c21c0.mov

OLFDB committed:
WINCE build issue fixed

OLFDB committed:
Code source added

OLFDB committed:
Cleanup and sanitycheck

jkoan commented on navit/osd/core/osd_core.c:
```diff
@@ -2239,7 +2239,10 @@ static void osd_nav_toggle_announcer_draw(struct osd_priv_common *opc, struct na
         if (!gr_image) {
             g_free(path);
-            path = graphics_icon_path("unknown.png");
+            if (this->active)
+                path = graphics_icon_path("sound_on.png");
+            else
+                path = graphics_icon_path("sound_off.png");
```
Strictly speaking this dosn't belong into this PR

OLFDB replied:
You are right, do I need to remove it?

jkoan commented on navit/maptool/osm.c:
```diff
@@ -139,800 +162,840 @@ struct country_table {
+
+static char* cleanCondition(char* condition) {
+
+    while(*condition==32) { //remove leading whitespace TODO: do it in maptool
```
We are already within maptool

OLFDB replied:
Removed the comment.

jkoan commented on navit/maptool/osm.c:
```diff
@@ -139,800 +162,840 @@ struct country_table {
+    while(*condition==32) { //remove leading whitespace TODO: do it in maptool
+        condition++;
+    }
+    while(*(condition+strlen(condition)-1)==32) { //remove trailing whitespace TODO: do it in maptool
```
This as well

jkoan commented on navit/osd/core/osd_core.c:
```diff
@@ -2857,6 +2862,455 @@ static struct osd_priv *osd_speed_warner_new(struct navit *nav, struct osd_metho
+    while (ptrcond = strstr(condition, "(")) {
+        *ptrcond = 32;
+    }
+    while(*condition==32) { //remove leading whitespace TODO: do it in maptool
```
Is this still current? see comment in maptool

jkoan commented on navit/osd/core/osd_core.c:
```diff
@@ -2857,6 +2862,455 @@ static struct osd_priv *osd_speed_warner_new(struct navit *nav, struct osd_metho
+    while(*condition==32) { //remove leading whitespace TODO: do it in maptool
+        condition++;
+    }
+    while(*(condition+strlen(condition)-1)==32) { //remove trailing whitespace TODO: do it in maptool
```
as well

OLFDB replied:
Removed the comment.

jkoan commented on navit/route_protected.h:
```diff
@@ -130,32 +155,44 @@ struct route_graph_segment {
 	                                         *  same point. Start of this list is in route_graph_point->end. */
 	struct route_graph_point *start;		/**< Pointer to the point this segment starts at. */
 	struct route_graph_point *end;			/**< Pointer to the point this segment ends at. */
+	struct route_graph_lez *lez;
 	struct route_segment_data data;			/**< The segment data */
 };
 
+
+
+struct route_graph_lezs {
+    struct route_graph_lez *next;    /**< Pointer to the next lez */
+    unsigned nlezs;          /**< How many lezs do we have? */
+    struct route_graph_lez lez[0];     /**< Pointer to the lez structure */
+    /* WARNING: There will be lezs following here, so do not create new fields after c! */
+};
+
+
+
 /**
  * @brief A complete route graph
  *
  * The route graph holds all routable segments along with the connections between them and the cost of
  * each segment.
  */
 struct route_graph {
-	int busy;                                   /**< Route calculation is in progress: the graph is being built,
-	                                             *   flooded or the path is being built (a more detailed status can be
-	                                             *   obtained from the route’s status attribute) */
+	int busy;                                   /**< The graph is being built */
```
Why removing a longer more explaining comment? Is is outdated?

jkoan commented on navit/route_protected.h:
```diff
@@ -130,32 +155,44 @@ struct route_graph_segment {
+	struct callback *donelez_cb;                /**< Callback when graph is done */
 	struct event_idle *idle_ev;                 /**< The pointer to the idle event */
 	struct route_graph_segment *route_segments; /**< Pointer to the first route_graph_segment in the linked list of all segments */
-	struct route_graph_segment *avoid_seg;      /**< Segment to which a turnaround penalty (if active) applies */
```
Why removing this comment?

jkoan commented on navit/vehicleprofile.h:
```diff
@@ -29,6 +31,39 @@ enum maxspeed_handling {
     maxspeed_ignore = 2,		/*!< Ignore maxspeed of segment, always use {@code route_weight} of road profile */
 };
 
+enum emissionclass {
+    //Truck
+    Euro_0,
+    Euro_1,
+    Euro_2,
+    Euro_3,
+    Euro_4,
+    Euro_5,
+    Euro_6,
+    //Car, use label colors and Euro_6d_TEMP for now
+    Cat_RED,
+    Cat_YLW,
+    Cat_GRN,
+    Euro_6d_TEMP,
+    EMISSION_CLASS_MAX,
+};
+
+static struct emissionclasstxt {
+  char  *classes[EMISSION_CLASS_MAX];
+} ectxt = {
+            "Euro 0",
+            "Euro 1",
+            "Euro 2",
+            "Euro 3",
+            "Euro 4",
+            "Euro 5",
+            "Euro 6",
+            "Red Label",
+            "Yellow Label",
+            "Green Label",
+            "Euro 6d temp"
+};
+
```
This seems to be really specific to Germany, what about other country's?

OLFDB replied:
The Euro x classes are valid in EU, not only in Germany. For the labels you are right. 
The purpose of the emission class was to use it for routing with the low emission zones. As the access tag for the lezs is missing often, it is not usable at the moment. That's why I added the "Enter low emission zone" value that is used for routing now.

jkoan commented on scripts/update_download_center.sh:
```diff
@@ -35,7 +35,7 @@ echo "Setting up all Variables"
 TMP_DIR=$(mktemp -d)
 CIRCLECI_API_BASE="https://circleci.com/api/v1.1/"
-NAVIT_DOWNLOAD_CENTER_REPO="git@github.com:navit-gps/download-center"
+NAVIT_DOWNLOAD_CENTER_REPO="git@github.com:OLFDB/download-center"
```
please revert this

OLFDB replied:
Same as above. Would like to keep it in my fork only, so my CI build is working. Is that somehow possible?

jkoan replied:
Need to revert d670c8ca19fa8976e389f6595420a29a810e9ed4 for this

jkoan reviewed (changes requested):
In addition to the other comments i noticed that that your branch needs to be rebased, are you able to do so?

OLFDB committed:
Revert "Update config.yml"

This reverts commit e34ee3bb8ed982dd658cc36d8753044d5c446d6b.

OLFDB commented on navit/attr.h:
```diff
@@ -65,7 +65,7 @@ enum attr_format {
 #define AF_ROUNDABOUT_VALID	(1<<4)
 #define AF_ONEWAY_EXCEPTION	(1<<5)
 #define AF_SPEED_LIMIT		(1<<6)
-#define AF_RESERVED1		(1<<7)
```
A grep for AF_RESERVED1 didn't show any usage.

OLFDB committed:
apply 2146fcd9c9ce1560b725e481dc8b808dd425b1d2

OLFDB committed:
Removed text from image

OLFDB commented on navit/icons/stvo_1053_35.svg:
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Removed text from image.

OLFDB committed:
Removed todo comments

OLFDB committed:
Merge branch 'trunk' into trucknavigation

OLFDB committed:
Merge branch 'trucknavigation' of https://github.com/OLFDB/navit into trucknavigation

OLFDB committed:
Merge branch 'trunk' of https://github.com/navit-gps/navit into navit-gps-trunk

OLFDB committed:
Merge branch 'navit-gps-trunk' into trucknavigation

OLFDB committed:
to trunk

OLFDB committed:
do not parse lez when lez is allowed in vehicleprofile to improve performance when calculating the route

OLFDB commented:
> Good work. I'm looking forward for the low emmission zones, as those can be very interesting on map.

@metalstrolch: When testing with larger maps having more than one low emission zone, the detection of the lezs is not working properly. I think I need somehow to have a reference to the tiles containing the lez to be able to select the correct polygon to test against during route calculation. Could you help me with that?

OLFDB committed:
Exclude eclispe _build directory

OLFDB committed:
added country_set_default_country

OLFDB committed:
added navit_store_center

DRAFT: route_set_position added parameter gra

OLFDB committed:
added country_set_default_country

OLFDB committed:
handle conditional speed limits

DRAFT: get route in conditional speed warner

OLFDB committed:
handle conditional speed limit

OLFDB committed:
Format and additional data for debugging LEZ routing

OLFDB committed:
route_set_position added graphics parameter

OLFDB committed:
added min and max to struct route_graph_lez

OLFDB committed:
Address error: format not a string literal and no format arguments
[-Werror=format-security]
